### PR TITLE
enclave-tls: fix an issue about segment fault when running sgx ecdsa …

### PR DESCRIPTION
**Pull request by YangLiang3**
…qve instance

Fixes: #980
Signed-off-by: Liang Yang <liang3.yang@intel.com>

### CODE REVIEW

File: enclave-tls/src/tls_wrappers/api/tls_wrapper_verify_certificate_extension.c [hunk 1]
```diff
@@ -16,24 +16,16 @@ tls_wrapper_err_t tls_wrapper_verify_certificate_extension(tls_wrapper_ctx_t *tl
...
-
-		//quote_ctx->quote_private = tls_ctx->tls_private;
-		enclave_quote_err_t err =
-			quote_ctx->opts->verify_evidence(quote_ctx, evidence,
-							 hash, hash_len);
-		if (err != ENCLAVE_QUOTE_ERR_NONE) {
-			ETLS_ERR("failed to verify evidence %#x\n", err);
-			return err;
-		}
+	if (!(tls_ctx) || !(tls_ctx->etls_handle) || !(tls_ctx->etls_handle->verifier) || !(tls_ctx->etls_handle->verifier->opts) ||
```
Comment (jiazhang0) [inline enclave-tls/src/tls_wrappers/api/tls_wrapper_verify_certificate_extension.c:19]: Dot't need to use `()`.

File: enclave-tls/src/include/enclave-tls/tls_wrapper.h [hunk 2]
```diff
@@ -45,6 +45,7 @@ typedef struct {
 } tls_wrapper_opts_t;
 
 struct tls_wrapper_ctx {
+	enclave_tls_handle etls_handle;
```
Comment (jiazhang0) [inline enclave-tls/src/include/enclave-tls/tls_wrapper.h:48]: Should be `etls_core_context_t *`, and add some comments about why adding this field.

File: enclave-tls/src/tls_wrappers/api/tls_wrapper_verify_certificate_extension.c [hunk 3]
```diff
@@ -16,24 +16,16 @@ tls_wrapper_err_t tls_wrapper_verify_certificate_extension(tls_wrapper_ctx_t *tl
...
+	    !(tls_ctx->etls_handle->verifier->opts->verify_evidence))
+		return -TLS_WRAPPER_ERR_INVALID;
+
+	enclave_quote_err_t err =
+		tls_ctx->etls_handle->verifier->opts->verify_evidence(tls_ctx->etls_handle->verifier, evidence,
```
Comment (jiazhang0) [inline enclave-tls/src/tls_wrappers/api/tls_wrapper_verify_certificate_extension.c:24]: Verifier should be determined by evidence type. You are choosing to use a fixed verifier instance to cover all possible retrieved evidence. This is unexpected.

So do you want to implement the dynamic verifier determination in later commit?

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: enclave-tls/src/include/enclave-tls/tls_wrapper.h (hunk 2 above)
Comment (YangLiang3) [reply enclave-tls/src/include/enclave-tls/tls_wrapper.h:48]: enclave_tls_handle belongs to api.h which can be exposed to user.

File: enclave-tls/src/tls_wrappers/api/tls_wrapper_verify_certificate_extension.c (hunk 3 above)
Comment (YangLiang3) [reply enclave-tls/src/tls_wrappers/api/tls_wrapper_verify_certificate_extension.c:29]: currently judge evidence type and verifier type. If it's not consistent,  I will return err and message some error info.

File: enclave-tls/src/api/enclave_tls_init.c [hunk 4]
```diff
@@ -58,14 +58,25 @@ enclave_tls_err_t enclave_tls_init(const enclave_tls_conf_t *conf,
 	if (err != ENCLAVE_TLS_ERR_NONE)
 		goto err_ctx;
 
-	/* Select the target enclave quote to be used */
+	/* Select the target enclave quote attester to be used */
```
Comment (jiazhang0) [inline enclave-tls/src/api/enclave_tls_init.c:61]: Change to `Select the target attester to be used`.

File: enclave-tls/src/api/enclave_tls_init.c [hunk 5]
```diff
@@ -58,14 +58,25 @@ enclave_tls_err_t enclave_tls_init(const enclave_tls_conf_t *conf,
...
 	choice = ctx->config.attester_type;
 	if (choice[0] == '\0') {
 		choice = global_core_context.config.attester_type;
 		if (choice[0] == '\0')
 			choice = NULL;
 	}
-	err = etls_enclave_quote_select(ctx, choice, ctx->config.cert_algo);
+	err = etls_enclave_quote_attester_select(ctx, choice, ctx->config.cert_algo);
```
Comment (jiazhang0) [inline enclave-tls/src/api/enclave_tls_init.c:68]: Change to `etls_attester_select`. `etls_enclave_quote_attester_select` is somewhat redundancy.

File: enclave-tls/src/api/enclave_tls_init.c [hunk 6]
```diff
@@ -58,14 +58,25 @@ enclave_tls_err_t enclave_tls_init(const enclave_tls_conf_t *conf,
...
+	if (err != ENCLAVE_TLS_ERR_NONE)
+		goto err_ctx;
+
+        /* Select the target enclave quote verifier to be used */
```
Comment (jiazhang0) [inline enclave-tls/src/api/enclave_tls_init.c:72]: fix the indent.

File: enclave-tls/src/api/enclave_tls_init.c [hunk 7]
```diff
@@ -58,14 +58,25 @@ enclave_tls_err_t enclave_tls_init(const enclave_tls_conf_t *conf,
...
+	choice = ctx->config.verifier_type;
+	if (choice[0] == '\0') {
+		choice = global_core_context.config.verifier_type;
+		if (choice[0] == '\0')
+			choice = NULL;
+	}
+	err = etls_enclave_quote_verifier_select(ctx, choice, ctx->config.cert_algo);
```
Comment (jiazhang0) [inline enclave-tls/src/api/enclave_tls_init.c:79]: change to `etls_verifier_select`

File: enclave-tls/src/enclave_quotes/internal/etls_enclave_quote_select.c [hunk 8]
```diff
@@ -63,10 +63,56 @@ enclave_tls_err_t etls_enclave_quote_select(etls_core_context_t *ctx,
 		return -ENCLAVE_TLS_ERR_INVALID;
 	}
 
-	ctx->attester = ctx->verifier = quote_ctx;
+	ctx->attester = quote_ctx;
 	ctx->flags |= ENCLAVE_TLS_CTX_FLAGS_QUOTING_INITIALIZED;
 
-	ETLS_INFO("the enclave quote '%s' selected\n", ctx->attester->opts->type);
+	ETLS_INFO("the enclave quote attester'%s' selected\n", ctx->attester->opts->type);
```
Comment (jiazhang0) [inline enclave-tls/src/enclave_quotes/internal/etls_enclave_quote_select.c:69]: change to `the attester '%s' selected`

File: enclave-tls/src/enclave_quotes/internal/etls_enclave_quote_select.c [hunk 9]
```diff
@@ -63,10 +63,56 @@ enclave_tls_err_t etls_enclave_quote_select(etls_core_context_t *ctx,
...
+
+	return ENCLAVE_TLS_ERR_NONE;
+}
+
+enclave_tls_err_t etls_enclave_quote_verifier_select(etls_core_context_t *ctx,
```
Comment (jiazhang0) [inline enclave-tls/src/enclave_quotes/internal/etls_enclave_quote_select.c:74]: change to `etls_verifier_select()`

File: enclave-tls/src/enclave_quotes/internal/etls_enclave_quote_select.c [hunk 10]
```diff
@@ -63,10 +63,56 @@ enclave_tls_err_t etls_enclave_quote_select(etls_core_context_t *ctx,
...
+					    const char *type,
+					    enclave_tls_cert_algo_t algo)
+{
+	ETLS_DEBUG("selecting the enclave quote '%s' ...\n", type);
```
Comment (jiazhang0) [inline enclave-tls/src/enclave_quotes/internal/etls_enclave_quote_select.c:78]: change to `selecting the verifier '%s' ...`

File: enclave-tls/src/enclave_quotes/internal/etls_enclave_quote_select.c [hunk 11]
```diff
@@ -63,10 +63,56 @@ enclave_tls_err_t etls_enclave_quote_select(etls_core_context_t *ctx,
...
+			ETLS_ERR("failed to select an enclave quote\n");
+		else
+			ETLS_ERR("failed to select the enclave quote '%s'\n", type);
+
+		return -ENCLAVE_TLS_ERR_INVALID;
+	}
+
+	ctx->verifier = quote_ctx;
+
+	ETLS_INFO("the enclave quote verifier'%s' selected\n", ctx->verifier->opts->type);
```
Comment (jiazhang0) [inline enclave-tls/src/enclave_quotes/internal/etls_enclave_quote_select.c:115]: ditto

File: enclave-tls/src/include/enclave-tls/tls_wrapper.h [hunk 12]
```diff
@@ -45,6 +45,8 @@ typedef struct {
 } tls_wrapper_opts_t;
 
 struct tls_wrapper_ctx {
+        /* associate tls wrapper with enclave quote's verifier and attester */
```
Comment (jiazhang0) [inline enclave-tls/src/include/enclave-tls/tls_wrapper.h:48]: fix the indent

File: enclave-tls/src/include/enclave-tls/tls_wrapper.h (hunk 2 above)
Comment (jiazhang0) [reply enclave-tls/src/include/enclave-tls/tls_wrapper.h:49]: OK.

File: enclave-tls/src/include/internal/enclave_quote.h [hunk 13]
```diff
@@ -13,10 +13,12 @@
 
 extern enclave_tls_err_t etls_enclave_quote_load_all(void);
 extern enclave_tls_err_t etls_enclave_quote_load_single(const char *);
-extern enclave_tls_err_t etls_enclave_quote_select(etls_core_context_t *,
+extern enclave_tls_err_t etls_enclave_quote_attester_select(etls_core_context_t *,
+						   const char *,
+						   enclave_tls_cert_algo_t);
+extern enclave_tls_err_t etls_enclave_quote_verifier_select(etls_core_context_t *,
```
Comment (jiazhang0) [inline enclave-tls/src/include/internal/enclave_quote.h:19]: change the names.

File: enclave-tls/src/tls_wrappers/api/tls_wrapper_verify_certificate_extension.c [hunk 14]
```diff
@@ -16,24 +16,21 @@ tls_wrapper_err_t tls_wrapper_verify_certificate_extension(tls_wrapper_ctx_t *tl
...
-
-		//quote_ctx->quote_private = tls_ctx->tls_private;
-		enclave_quote_err_t err =
-			quote_ctx->opts->verify_evidence(quote_ctx, evidence,
-							 hash, hash_len);
-		if (err != ENCLAVE_QUOTE_ERR_NONE) {
-			ETLS_ERR("failed to verify evidence %#x\n", err);
-			return err;
-		}
+	if (!tls_ctx || !(tls_ctx->etls_handle) || !(tls_ctx->etls_handle->verifier) || !(tls_ctx->etls_handle->verifier->opts) ||
```
Comment (jiazhang0) [inline enclave-tls/src/tls_wrappers/api/tls_wrapper_verify_certificate_extension.c:19]: don't use () for ! check.

File: enclave-tls/src/tls_wrappers/wolfssl/un_negotiate.c [hunk 15]
```diff
@@ -193,9 +193,9 @@ static crypto_wrapper_err_t calc_pubkey_hash(DecodedCert *crt,
 }
 
 #ifdef WOLFSSL_SGX_WRAPPER
-int verify_certificate(uint8_t *der_cert, uint32_t der_cert_len)
+int verify_certificate(void *userCtx, uint8_t *der_cert, uint32_t der_cert_len)
 {
-	tls_wrapper_ctx_t *tls_ctx;
+	tls_wrapper_ctx_t *tls_ctx = userCtx;
 #else
 int verify_certificate(int preverify, WOLFSSL_X509_STORE_CTX *store)
```
Comment (jiazhang0) [inline enclave-tls/src/tls_wrappers/wolfssl/un_negotiate.c:200]: Is there way to unify the prototype of verify_certificate() in another commit?

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: enclave-tls/src/tls_wrappers/wolfssl/un_negotiate.c (hunk 15 above)
Comment (YangLiang3) [reply enclave-tls/src/tls_wrappers/wolfssl/un_negotiate.c:200]: I will check, if it can be unified, I will change

**Review by Ruoyu-y — APPROVED** [review]
LGTM

File: enclave-tls/src/tls_wrappers/wolfssl/un_negotiate.c (hunk 15 above)
Comment (YangLiang3) [reply enclave-tls/src/tls_wrappers/wolfssl/un_negotiate.c:200]: if unifying the prototype, then WOLFSSL_X509_STORE_CTX structure will exposed to unstrust part completely.

File: enclave-tls/src/enclave_quotes/internal/etls_enclave_quote_select.c [hunk 16]
```diff
@@ -56,17 +56,63 @@ enclave_tls_err_t etls_enclave_quote_select(etls_core_context_t *ctx,
...
 
 		return -ENCLAVE_TLS_ERR_INVALID;
 	}
 
-	ctx->attester = ctx->verifier = quote_ctx;
+	ctx->attester = quote_ctx;
 	ctx->flags |= ENCLAVE_TLS_CTX_FLAGS_QUOTING_INITIALIZED;
 
-	ETLS_INFO("the enclave quote '%s' selected\n", ctx->attester->opts->type);
+	ETLS_INFO("the attester'%s' selected\n", ctx->attester->opts->type);
```
Comment (jiazhang0) [inline enclave-tls/src/enclave_quotes/internal/etls_enclave_quote_select.c:69]: add the space between r and '.

File: enclave-tls/src/enclave_quotes/internal/etls_enclave_quote_select.c [hunk 17]
```diff
@@ -56,17 +56,63 @@ enclave_tls_err_t etls_enclave_quote_select(etls_core_context_t *ctx,
...
+			ETLS_ERR("failed to select a verifier\n");
+		else
+			ETLS_ERR("failed to select the verifier '%s'\n", type);
+
+		return -ENCLAVE_TLS_ERR_INVALID;
+	}
+
+	ctx->verifier = quote_ctx;
+
+	ETLS_INFO("the verifier'%s' selected\n", ctx->verifier->opts->type);
```
Comment (jiazhang0) [inline enclave-tls/src/enclave_quotes/internal/etls_enclave_quote_select.c:115]: ditto

File: enclave-tls/src/tls_wrappers/api/tls_wrapper_verify_certificate_extension.c [hunk 18]
```diff
@@ -11,30 +11,31 @@
...
+			!tls_ctx->etls_handle->verifier->opts->verify_evidence)
+		return -TLS_WRAPPER_ERR_INVALID;
 
-		if (strcmp(evidence->type, quote_ctx->opts->type))
-			continue;
-
-		if (!quote_ctx || !(quote_ctx->opts) ||
-		    !(quote_ctx->opts->verify_evidence))
+	if (strcmp(tls_ctx->etls_handle->verifier->opts->type, evidence->type)) {
+		ETLS_WARN("type doesn't match between verifier '%s' and attester '%s'\n", tls_ctx->etls_handle->verifier->opts->type, evidence->type);
```
Comment (jiazhang0) [inline enclave-tls/src/tls_wrappers/api/tls_wrapper_verify_certificate_extension.c:24]: change to `and evidence `.

File: enclave-tls/src/tls_wrappers/wolfssl-sgx/ocalls.c [hunk 19]
```diff
@@ -17,8 +17,8 @@ size_t ocall_send(int sockfd, const void *buf, size_t len, int flags)
         return send(sockfd, buf, len, flags);
 }
 
-int ocall_verify_certificate(uint8_t *der_crt, uint32_t der_crt_len)
+int ocall_verify_certificate(void *userCtx, uint8_t *der_crt, uint32_t der_crt_len)
```
Comment (jiazhang0) [inline enclave-tls/src/tls_wrappers/wolfssl-sgx/ocalls.c:20]: change to ctx.

File: enclave-tls/src/tls_wrappers/wolfssl/un_negotiate.c (hunk 15 above)
Comment (jiazhang0) [reply enclave-tls/src/tls_wrappers/wolfssl/un_negotiate.c:200]: OK

File: enclave-tls/src/tls_wrappers/wolfssl/un_negotiate.c [hunk 20]
```diff
@@ -193,9 +193,9 @@ static crypto_wrapper_err_t calc_pubkey_hash(DecodedCert *crt,
 }
 
 #ifdef WOLFSSL_SGX_WRAPPER
-int verify_certificate(uint8_t *der_cert, uint32_t der_cert_len)
+int verify_certificate(void *userCtx, uint8_t *der_cert, uint32_t der_cert_len)
```
Comment (jiazhang0) [inline enclave-tls/src/tls_wrappers/wolfssl/un_negotiate.c:196]: change to ctx.

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

**Review by haosanzi — APPROVED** [review]

**Pull request merged**